templates/terraform/encoders/disk.erb

```diff
@@ -5,7 +5,11 @@ if err != nil {
  	return nil, err
 }
 
+<% unless compiler == "terraformobjectlibrary-codegen" -%>
 userAgent, err := generateUserAgentString(d, config.userAgent)
+<% else %>
+userAgent, err := generateUserAgentString(d.(*schema.ResourceData), config.userAgent)
```

Comment:
Does `d` have a different type in conversion? We should probably fix that rather than generate different code here.

Reply:
@rileykarson the general difference between the provider code and the conversion code is that `d`  in the provider code is of type `*schema.ResourceData` which is a struct (https://github.com/hashicorp/terraform-plugin-sdk/blob/c31bf17bbcf68029d5174cf767c5c3e7ca3a09cd/helper/schema/resource_data.go#L22)

and `d` in the conversion is of type `TerraformResourceData` which is an interface https://github.com/GoogleCloudPlatform/magic-modules/blob/5a0c2673d646611f4932b835a37927d72973a3e8/third_party/terraform/utils/utils.go.erb#L23


For example, the generator code for conversion is

https://github.com/GoogleCloudPlatform/magic-modules/blob/5a0c2673d646611f4932b835a37927d72973a3e8/templates/terraform/objectlib/base.go.erb#L17

and the code for the provider is

https://github.com/GoogleCloudPlatform/magic-modules/blob/5a0c2673d646611f4932b835a37927d72973a3e8/templates/terraform/resource.erb#L147

Reply:
This type cast will just fail though, right?

Reply:
The original error in the conversion code without the conditional is:

```
terraform-google-conversion\google\compute_disk.go:331:43:  cannot use d (type TerraformResourceData) as type *"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema".ResourceData in argument to generateUserAgentString: need type assertion
```
which needs to use the cast `d.(*schema.ResourceData)`
 
but if we add the cast to the template we get this error in the provider code 

```
google/resource_compute_disk.go:1451:45: invalid type assertion: d.(*schema.ResourceData) (non-interface type *schema.ResourceData on left)
```

Reply:
Can we extend the `TerraformResourceData` interface to include the information we pull off the `*schema.ResourceData` type here, and use the interface as the expected type in `generateUserAgentString` instead?

Reply:
the information was added in a previous PR

https://github.com/GoogleCloudPlatform/magic-modules/blob/caac9b1e40ad9c0da650ebfecf61be78e3f07ada/third_party/terraform/utils/utils.go.erb#L31

I will test the suggested change to check the impact it will have

Reply:
Signature changed